iceoryx_posh/include/iceoryx_posh/popo/modern_api/base_subscriber.hpp

```diff
@@ -118,10 +118,23 @@ class BaseSubscriber
     ///            later by the user
     /// @return success if the subscriber is attached otherwise an WaitSetError enum which describes
     ///            the error
-    cxx::expected<WaitSetError> attachTo(WaitSet& waitset,
-                                         const SubscriberEvent subscriberEvent,
-                                         const uint64_t triggerId = Trigger::INVALID_TRIGGER_ID,
-                                         const Trigger::Callback<Subscriber> callback = nullptr) noexcept;
+    template <uint64_t WaitSetCapacity>
+    cxx::expected<WaitSetError> attachEvent(WaitSet<WaitSetCapacity>& waitset,
+                                            const SubscriberEvent subscriberEvent,
+                                            const uint64_t triggerId = Trigger::INVALID_TRIGGER_ID,
+                                            const Trigger::Callback<Subscriber> callback = nullptr) noexcept;
+
+    /// @brief attaches a WaitSet to the subscriber
+    /// @param[in] waitset reference to the waitset to which the subscriber should be attached to
+    /// @param[in] subscriberEvent the event which should be attached
+    /// @param[in] callback callback which is attached to the trigger and which can be called
+    ///            later by the user
+    /// @return success if the subscriber is attached otherwise an WaitSetError enum which describes
+    ///            the error
+    template <uint64_t WaitSetCapacity>
+    cxx::expected<WaitSetError> attachEvent(WaitSet<WaitSetCapacity>& waitset,
```

Comment:
with the risk to start some bikeshedding, this might be confusing when reading in code since the event is not attached to the corresponding object of the method, e.g. `mySubscriber.attachEvent(...)`. What it actually does is a `mySubscriber.attachEventToWaitset(...)`. Similarly with `detachEvent`.

From my point of view, a more natural way would be to add events to a `WaitSet`, e.g. `myWaitSet.addEvent(mySubscriber, SubscriberEvent::HAS_SAMPLES, ...)`. The `WaitSet` also does not need to know the `Subscriber` when `addEvent` is a template method.
```
template <typename T, typename E>
cxx::expected<WaitSetError> addEvent(T& eventSource, E& event, Trigger::Callback<T> callback) {
    return eventSource.attachEventToWaitset(*this, event, callback);
}
```
In this scenario, `attachEventToWaitset` could be `protected` and the `WaitSet` a friend class. This is only necessary if we want to restrict the number of options a avent can be attached to a `WaitSet`.

Reply:
fixed by redesign